File: gen.py
import re
import requests
from tqdm import tqdm
import sys

# fn = "https://peter.crall.family/minecraft/cc/pos/os/gui/Window.lua"
fn = "Window.lua"
if len(sys.argv) > 1:
    fn = sys.argv[1]
print("Reading", fn)
lines = []
if fn.startswith("http"):
    response = requests.get(fn)
    lines = [re.sub("^\s*", "", line) for line in str(response.text).split("\n")]
else:
    with open('Window.lua') as file:
        for line in file:
            lines.append(re.sub("^\s*", "", line[:-1]))

classname = ""
classdesc = ""
funcs = []
vars = []

def htmlType(type, nonil=False):
    if not nonil and (type == "nil" or type == "any"):
        return htmlNil(type)
    return f'<code class="type">{type}</code>'
def htmlVar(var):
    return f'<code class="var">{var}</code>'
def htmlNil(type="nil"):
    return f'<code class="nil">{type}</code>'

class Function():
    def __init__(self):
        self.name = ""
        self.type = ""
        self.desc = []
        self.args = []
        self.ret = []
    
    def __str__(self):
        s = f"{self.type:>8s} {self.name} ({', '.join([str(a) for a in self.args])})"
        for r in self.ret:
            s += f"\n -> {r}"
        for d in self.desc:
            s += f"\n{d}"
        return s
    
    def htmlSig(self):
        return f'<code class="func">{self.name}({", ".join([a.htmlSig() for a in self.args])})</code>'

    def htmlSummaryTableRow(self):
        s = f'<tr>\n<td>'
        if len(self.ret) > 0:
            s += ', '.join([htmlType(r.type) for r in self.ret])
        else:
            s+= htmlNil()
        s += '</td>\n'
        s += f'<td>{self.htmlSig()}</td>\n'
        s += f'<td>{"<br>".join(self.desc)}</td>\n'
        return s + '</tr>\n'

    def htmlDetailBlock(self):
        s = f'<div class="datablock" id={self.name}>'
        s += f'<h3>{self.htmlSig()}</h3>'
        s += f'{"<br/>".join(self.desc)}'
        if len(self.args) > 0:
            s += "<br/><u>Args</u><br/>"
            s += "<br/>".join([f"{htmlVar(arg.name)}: {htmlType(arg.type)} {'- ' if len(arg.desc) > 0 else ''}{arg.desc}" for arg in self.args])
        if len(self.ret) > 0:
            s += "<br/><u>Returns</u><br/>"
            s += "<br/>".join([f"{htmlVar(r.name)}: {htmlType(r.type)} {'- ' if len(r.desc) > 0 else ''}{r.desc}" for r in self.ret])
        return s + "</div>"

class Variable():
    def __init__(self, name="", type="", desc="", val=""):
        self.name = name
        self.type = type
        self.desc = desc
        self.val = val

    def __str__(self):
        return f"{self.name}{'=' if len(self.val)>0 else ''}{self.val}: {self.type}; {self.desc}"
    
    def htmlSig(self):
        return htmlVar(self.name)+": "+htmlType(self.type)
    
    def htmlSummaryTableRow(self):
        return f'<tr><td>{htmlType(self.type)}</td><td>{htmlVar(self.name)}</td><td>{self.desc}</td></tr>'
    
    def htmlDetailBlock(self):
        return f'<div class="datablock" id={self.name}><h3>{self.htmlSig()}</h3>{self.desc}<br/>Default: {self.val}</div>'

class Argument(Variable):
    pass

def split(regex, str):
    s = re.search(regex, str)
    if s is not None:
        sp = s.span()
        return [str[:sp[0]], str[sp[1]:]]
    else:
        return [str, ""]

print("Parsing ...")
for i in range(len(lines)):
    l = lines[i]
    if l.startswith('---@class '):
        classname = l.split(" ")[1]
        classdesc = " ".join(l.split(" ")[2:])
        for j in range(i, len(lines)):
            l = lines[j]
            if len(l) > 0:
                if l.startswith('---'):
                    if l.startswith('---@field'):
                        v = Variable()
                        l = split("---@field\s*", l)[1]
                        s = split("\s+", l)
                        v.name = s[0]
                        s = split("\s+", s[1])
                        v.type = s[0]
                        v.desc = s[1]
                        vars.append(v)
                else:
                    break
    elif classname != "" and l.startswith('local ' + classname):
        for j in range(i+1, len(lines)):
            l = lines[j]
            if l.startswith("}"):
                break
            if len(l) > 0:
                v = Variable()
                s = split("\s*=\s*", l)
                v.name = s[0]
                s = split("\s*---\s*", s[1])
                v.val = s[0].replace(",", "")
                v.type = "any"
                v.desc = s[1]
                for var in vars:
                    if var.name == v.name:
                        var.val = v.val
                        v = None
                        break
                if v is not None:
                    vars.append(v)
    elif l.startswith("function"):
        f = Function()
        # get the function signature
        f.name = l[9:].split("(")[0]
        f.type = "func"
        if(f.name.startswith(classname)):
            f.type = "instance" if f.name[len(classname)] == ":" else "static"
            f.name = f.name[len(classname)+1:]
        for arg in re.search("\(.*\)", l)[0][1:-1].split(","):
            if len(arg) > 0:
                arg = split('^\s*', arg)[1]
                f.args.append(Argument(arg))
        # Find top of doc comment
        topline = i
        for j in range(i-1, 0, -1):
            if len(lines[j]) > 0:
                if not lines[j].startswith("---"):
                    topline = j+1
                    break
        if topline < i:
            for j in range(topline, i):
                l = lines[j]
                # print(l)
                if len(l) > 0:
                    if not l.startswith("---"):
                        print("EXPLOSION!", j, l)
                        exit()
                    l = split("^---\s*", l)[1]
                    if l[0] == "@":
                        if l.startswith("@param"):
                            sp = split("\s+", split("@param\s+", l)[1])
                            an = sp[0]
                            sp = split("\s+", sp[1])
                            at = sp[0]
                            ad = sp[1]
                            for a in f.args:
                                if a.name == an:
                                    a.type = at
                                    a.desc = ad
                                    break
                        elif l.startswith("@return"):
                            sp = split("\s+", split("@return\s+", l)[1])
                            rt = sp[0]
                            sp = split("\s+", sp[1])
                            rn = sp[0]
                            rd = sp[1]
                            r = Variable(rn, rt, rd)
                            f.ret.append(r)
                    else:
                        # print("De")
                        f.desc.append(l)
            
        funcs.append(f)

def plural(num, str):
    if num != 0:
        return f'{num} {str}'
    else:
        return f'{num} {str}s'

print(f"Found {plural(len(vars), 'variables')} and {plural(len(funcs), 'functions')}")

print("Downloading template")
response = requests.get("https://peter.crall.family/minecraft/cc/doc/classDef.html")
html = str(response.text)

import platform
if platform.node != "cheddarserver":
    html = html.replace('<link rel="stylesheet" href="/', '<link rel="stylesheet" href="https://peter.crall.family/')

html = html.replace("%ClassName%", classname)
html = html.replace("%ClassDesc%", classdesc)

print("Generating")
vsp = '\n'.join([v.htmlSummaryTableRow() for v in vars if not v.name.startswith("_")])
vst = '\n'.join([v.htmlSummaryTableRow() for v in vars if v.name.startswith("_") and not v.name.startswith("__")])
vsr = '\n'.join([v.htmlSummaryTableRow() for v in vars if v.name.startswith("__")])
vdp = '\n'.join([v.htmlDetailBlock() for v in vars if not v.name.startswith("_")])
vdt = '\n'.join([v.htmlDetailBlock() for v in vars if v.name.startswith("_") and not v.name.startswith("__")])
vdr = '\n'.join([v.htmlDetailBlock() for v in vars if v.name.startswith("__")])

html = html.replace("%TableVariables%", vsp + vst + vsr)

html = html.replace("%Variables%", vdp + vdt + vdr)

tifs = ""
tnifs = ""
ifdb = ""
nifdb = ""
for f in funcs:
    if f.type == "instance":
        tifs += f.htmlSummaryTableRow()
        ifdb += f.htmlDetailBlock()
    else:
        tnifs += f.htmlSummaryTableRow()
        nifdb += f.htmlDetailBlock()

html = html.replace("%TableFunctions%", tnifs)
html = html.replace("%TableInstanceFunctions%", tifs)

html = html.replace("%InstanceFunctions%", ifdb)
html = html.replace("%Functions%", nifdb)

print("Done")
with open("output.html", "w") as hf:
    hf.write(html)

File: gen.js
const fs = require('fs');
const { constrainedMemory } = require('process');

/*
 * Classes to store data in
 */
function getElemByName(list, name) {
    for(let i = 0; i < list.length; i++) {
        if(list[i].name == name) {
            return list[i];
        }
    }
    return undefined;
}

class Function {
    constructor() {
        this.name = "";
        this.type = "";
        this.desc = [];
        this.args = [];
        this.ret = [];
    }

    getArg(name) {
        return getElemByName(this.args, name);
    }

    getRet(name) {
        return getElemByName(this.ret, name);
    }

    getArgNames() {
        let s = [];
        this.args.forEach((a) => {
            s.push(a.name);
        });
        return s;
    }

    getRetTypes() {
        let s = [];
        this.ret.forEach((a) => {
            s.push(a.type);
        });
        return s;
    }

    toString() {
        let s = `${(this.type.length > 0) ? this.type + ' ' : ''}${this.name}(${this.getArgNames().join(', ')}) -> ${(this.ret.length > 0) ? this.getRetTypes().join("|") : "nil"}`
        this.desc.forEach((d) => {
            s += "\n\t" + d;
        });
        this.ret.forEach((r) => {
            s += `\n\tReturns ${r.name}:${r.type}${(r.desc.length > 0) ? ': ' + r.desc : ''}`;
        });
        return s
    }
}

class Variable {
    constructor(name, type, desc, val, vis) {
        this.name = name || "";
        this.type = type || "";
        this.desc = desc || "";
        this.val = val || "";
        this.vis = vis || "public";
        // console.log(this.vis, vis)
    }

    toString() {
        return `${this.name}${(this.val.length > 0) ? '=' : ''}${this.val}: ${this.type}(${this.vis}): ${this.desc}`
    }
}

class Argument extends Variable { }

// Info about the source data file
class LuaClass {
    constructor(name, parent, desc, vars, funcs) {
        this.name = name || "";
        this.parent = parent || "";
        this.desc = desc || [];
        this.vars = vars || [];
        this.funcs = funcs || [];
        this.const = undefined
    }

    getVar(name) {
        return getElemByName(this.vars, name);
    }

    getFunc(name) {
        return getElemByName(this.funcs, name);
    }
}

/*
 * Load the source file
 */

// TODO Should probably get these from args
let sourcefilename = "Window.lua"
let templatename = "classDef.html"
let outputfilename = "output.html"

// Load the code to parse
function loadFile(path) {
    console.log("Reading")
    const data = fs.readFileSync(path, 'utf8');
    return data;
}

function parse(lua) {
    console.log("Parsing");

    lines = lua.split(/\n/);
    lines.forEach((e, i, a) => {
        a[i] = e.split(/^\s*(.*)\s*$/)[1];
    });

    let classes = [];
    let funcs = [];

    for(let linenum = 0; linenum < lines.length; linenum++) {
        l = lines[linenum];
        if(l.startsWith('---@class ')) {
            let s = l.split(/^---@class\s+([^\s:]+)(?:\s*:\s*(\S+))?\s+(.*)$/);
            c = new LuaClass(s[1], s[2], [s[3]]);
            classes.push(c);
            for(let j = linenum; j < lines.length; j++) {
                ln = lines[j]
                if(ln.length > 0) {
                    if(ln.startsWith('---')) {
                        if(ln.startsWith('---@field')) {
                            let s = ln.split(/^---@field\s+(?:(private|protected|package)\s+)?(\S*)\s+(\S*)(?:\s+(.*))?\s*$/);
                            if (s.length > 4) {
                                var variable = new Variable(s[2], s[3], s[4], undefined, s[1]);
                                c.vars.push(variable);
                            }
                        } else if(ln.startsWith('---@class')) {
                            // Already used this
                        } else {
                            c.desc.push(ln.replace(/^---\s*/, ''));
                        }
                    } else {
                        break;
                    }
                }
            }
        } else {
        let s = l.split(/^local\s+(\S+)\s*=\s*{/);
        if(s.length > 1 && getElemByName(classes, s[1]) != undefined) {
            let c = getElemByName(classes, s[1]);
            for(let j = linenum+1; j < lines.length; j++) {
                l = lines[j];
                if(l.startsWith('}')) {
                    break;
                }
                if(l.length > 0) {
                    let s = l.split(/(\S+)\s*=\s*([^,]*)/);
                    if(c.getVar(s[1]) != undefined) {
                        c.getVar(s[1]).val = s[2];
                    } else {
                        c.vars.push(new Variable(s[1], val=[2]));
                    }
                }
            }
        } else {
        let s = l.split(/^function\s+(?:(\S+):)?(\S+)\s*\((.+)?\)/);
        if(s.length > 4) {
            let f = new Function();
            f.name = s[2];
            (s[3] || "").split(/,\s+/).forEach((a) => {
                f.args.push(new Argument(a));
            });
            if(getElemByName(classes, s[1]) != undefined) {
                f.type = "instance";
                getElemByName(classes, s[1]).funcs.push(f);
            } else {
                funcs.push(f);
            }
            let topline = linenum;
            for(let j = linenum - 1; j > 0; j--) {
                if(lines[j].length > 0) {
                    if(!lines[j].startsWith('---')) {
                        topline = j+1;
                        break;
                    }
                }
            }
            if(topline < linenum) {
                for(let j = topline; j < linenum; j++) {
                    let ln = lines[j];
                    if(ln.length > 0) {
                        if(!ln.startsWith('---')) {
                            console.error(j, ln);
                            throw Error("A line should have started with dashes but didn't!");
                        }
                        if(ln.startsWith('---@param')) {
                            let s = ln.split(/^---@param\s+(?:(private|protected|package)\s+)?(\S*)\s+(\S*)(?:\s+(.*))?\s*$/)
                            if(f.getArg(s[2]) != undefined) {
                                f.getArg(s[2]).type = s[3];
                                f.getArg(s[2]).desc = s[4];
                                f.getArg(s[2]).vis = s[1];
                            } else {
                                f.args.push(new Argument(s[2], s[3], s[4], s[1]));
                            }
                        } else if(ln.startsWith('---@return')) {
                            // not sure why this be here
                            /**@todo Fix this so the name of the variable is not "return"*/
                            let s = ln.split(/^---@return\s+(\S*)\s+(\S+)(?:\s+(.*))?\s*$/)
                            f.ret.push(new Variable(s[2], s[1], s[3]));
                        } else if (ln.startsWith('---@constructor')) {
                            // Mark this as the constructor for the named class
                            let s = ln.split(/^---@constructor\s+/)
                            getElemByName(classes, s[1]).const = f
                            // funcs.pop()
                            var index = funcs.indexOf(f);
                            if (index !== -1) {
                                funcs.splice(index, 1);
                            }
                        } else {
                            f.desc.push(ln.replace(/^---\s*/, ''));
                        }
                    }
                }
            }
        }}}
    }

    return [classes, funcs]
}


// this.name = "";
// this.desc = "";
// this.vars = [];
// this.funcs = [];

const toConsole = false

if (toConsole) {
    let r = parse(loadFile(sourcefilename))

    Object.keys(r[0]).forEach((k) => {
        c = r[0][k];
        console.log("Class " + c.name);
        c.desc.forEach((d) => {
            console.log('\t' + d);
        });
        console.log("Fields:");
        c.vars.forEach((k) => {
            console.log(k.toString('\t'));
        });
        console.log("Methods:");
        c.funcs.forEach((k) => {
            console.log(k.toString('\t'));
        });
    });

    console.log("Functions");
    r[1].forEach((f) => {
        console.log(f.toString());
    });

    console.log("done");
}


// l = '---@field private x number Window X origin'
// l = '---@field protected x number Window X origin'
// l = '---@field package x number Window X origin'
// l = '---@field x number Window X origin'
// l = '---@field x number'
// // private protected package
// let s = l.split(/^---@field\s+(?:(private|protected|package)\s+)?(\S*)\s+(\S*)(?:\s+(.*))?\s*$/)
// console.log(s)

module.exports = {
    loadFile,
    parse,
    Variable,
    Argument,
    Function,
    LuaClass
}